File: sdk/js/tests/unit/sensitive.test.ts
import { describe, expect, it } from "vitest";
import {
  array, decrypt, encrypt, importSchema, nullable, object, safeParseEncrypted, string,
} from "../../src/index.js";

const sensitive = <T extends ReturnType<typeof string>>(schema: T): T =>
  schema.describe("secret", { sensitive: true });

describe("sensitive data", () => {
  const schema = object({
    name: string(),
    secret: sensitive(string().minLength(3)),
    profile: object({ token: string() }).describe("profile", { sensitive: true }),
    aliases: array(sensitive(string())),
    note: nullable(sensitive(string())),
  });

  const plain = {
    name: "Ada",
    secret: "abc",
    profile: { token: "xyz" },
    aliases: ["one", "two"],
    note: null,
  };

  it("encrypts, validates, and decrypts sensitive leaves", () => {
    const seen: string[] = [];
    const encrypted = encrypt(schema, plain, (path, value) => {
      seen.push(path.join("."));
      return `encrypted:${JSON.stringify(value)}`;
    });

    expect(encrypted).toEqual({
      name: "Ada",
      secret: 'encrypted:"abc"',
      profile: 'encrypted:{"token":"xyz"}',
      aliases: ['encrypted:"one"', 'encrypted:"two"'],
      note: null,
    });
    expect(seen).toEqual(["secret", "profile", "aliases.0", "aliases.1"]);
    expect(safeParseEncrypted(schema, encrypted).success).toBe(true);
    expect(decrypt(schema, encrypted, (_path, value) =>
      JSON.parse((value as string).slice("encrypted:".length))
    )).toEqual(plain);
  });

  it("rejects broken envelopes and encrypt callbacks", () => {
    expect(safeParseEncrypted(schema, { ...plain, secret: "abc" }).success).toBe(false);
    expect(() => encrypt(schema, plain, () => "broken")).toThrow();
  });

  it("preserves null through sensitive refs to nullable targets", () => {
    const schema = importSchema({
      anyvaliVersion: "1.0", schemaVersion: "1.1",
      root: { kind: "ref", ref: "#/definitions/Secret", metadata: { sensitive: true } },
      definitions: { Secret: { kind: "nullable", inner: { kind: "string" } } }, extensions: {},
    } as any);
    const transform = () => { throw new Error("Null must not invoke the callback"); };
    expect(encrypt(schema, null, transform)).toBeNull();
    expect(safeParseEncrypted(schema, null)).toMatchObject({ success: true, data: null });
    expect(decrypt(schema, null, transform)).toBeNull();
  });
});

File: sdk/go/sensitive_test.go
package anyvali

import (
	"encoding/json"
	"testing"
)

func TestSensitiveDataRoundTripAndValidation(t *testing.T) {
	sensitive := DescribeOpts{Sensitive: true}
	schema := Object(map[string]Schema{
		"name":    String(),
		"secret":  String().Describe("secret", sensitive),
		"profile": Object(map[string]Schema{"token": String()}).Describe("profile", sensitive),
		"aliases": Array(String().Describe("alias", sensitive)),
	})
	plain := map[string]any{
		"name": "Ada", "secret": "abc",
		"profile": map[string]any{"token": "xyz"}, "aliases": []any{"one", "two"},
	}
	enc := func(_ []any, value any) (any, error) {
		b, _ := json.Marshal(value)
		return encryptedPrefix + string(b), nil
	}
	encrypted, err := Encrypt(schema, plain, enc)
	if err != nil {
		t.Fatal(err)
	}
	if !SafeParseEncrypted(schema, encrypted).Success {
		t.Fatal("encrypted value should validate")
	}
	decrypted, err := Decrypt(schema, encrypted, func(_ []any, value any) (any, error) {
		var out any
		err := json.Unmarshal([]byte(value.(string)[len(encryptedPrefix):]), &out)
		return out, err
	})
	if err != nil {
		t.Fatal(err)
	}
	if got, _ := json.Marshal(decrypted); string(got) != `{"aliases":["one","two"],"name":"Ada","profile":{"token":"xyz"},"secret":"abc"}` {
		t.Fatalf("unexpected round trip: %s", got)
	}
	if _, err := Encrypt(schema, plain, func([]any, any) (any, error) { return "broken", nil }); err == nil {
		t.Fatal("broken envelope should fail")
	}
}
